xdsl/parser.py

```diff
@@ -1640,21 +1641,65 @@ def not_implemented(_name: Span):
 
             return parsers.get(name.text, not_implemented)(name)
 
-    def _parse_builtin_dense_attr(self, _name: Span) -> Attribute | None:
-        err_msg = "Malformed dense attribute, format must be (`dense<` array-attr `>:` type)"  # noqa
-        self.parse_characters("<", err_msg)
-        info = list(self._parse_builtin_dense_attr_args())
-        self.parse_characters(">", err_msg)
-        self.parse_characters(":", err_msg)
+    def _parse_builtin_dense_attr(self,
+                                  _name: Span) -> DenseIntOrFPElementsAttr:
+        self._synchronize_lexer_and_tokenizer()
+        self.parse_punctuation('<', ' in dense attribute')
+
+        # The flatten list of elements
+        values: list[BaseParser._TensorLiteralElement]
+        # The dense shape.
+        # If it is `None`, then there is no values.
+        # If it is `[]`, then this is a splat attribute, meaning it has the same
+        # value everywhere.
+        shape: list[int] | None
+        if self._current_token.text == '>':
+            values, shape = [], None
+        else:
+            values, shape = self._parse_tensor_literal()
+        self.parse_punctuation('>', ' in dense attribute')
+
+        # Parse the dense type.
+        self.parse_punctuation(':', ' in dense attribute')
+        self._synchronize_lexer_and_tokenizer()
         type = self.expect(self.try_parse_type,
-                           "Dense attribute must be typed!")
+                           'Dense attribute must be typed!')
+        self._synchronize_lexer_and_tokenizer()
 
-        if not isa(type,
-                   RankedVectorOrTensorOf[AnyFloat | IntegerType | IndexType]):
+        # Check that the type is correct.
+        if not isa(
+                type, RankedVectorOrTensorOf[IntegerType]
+                | RankedVectorOrTensorOf[IndexType]
+                | RankedVectorOrTensorOf[AnyFloat]):
+            self.raise_error('Expected vector or tensor type of '
+                             'integer, index, or float type')
+
+        # Check that the shape matches the data when given a shaped data.
+        type_shape = [dim.value.data for dim in type.shape.data]
+        num_values = reduce((lambda x, y: x * y), type_shape, 1)
```

Comment:
supermeganit: there's a function called `prod` in `math` that does the same thing, and might be a little neater, if we're gaining an import anyway
```suggestion
        num_values = math.prod(type_shape)
```